File: lib/notification_service/notification_service/client/notification_client.py
# Copyright 2022 The AI Flow Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import abc
from datetime import datetime
from typing import List, Optional

from notification_service.model.event import Event


class ListenerRegistrationId(object):
    def __init__(self, id: str):
        self.id = id


class ListenerProcessor(object):
    def process(self, events: List[Event]):
        pass


class NotificationClient(metaclass=abc.ABCMeta):
    def __init__(self,
                 namespace: Optional[str] = None,
                 sender: Optional[str] = None):
        self.namespace = namespace
        self.sender = sender

    @abc.abstractmethod
    def send_event(self, event: Event):
        """
        Send event to Notification Sever.

        :param event: the event updated.
        :return: The created event.
        """
        pass

    @abc.abstractmethod
    def register_listener(self,
                          listener_processor: ListenerProcessor,
                          event_keys: List[str] = None,
                          offset: int = None
                          ) -> ListenerRegistrationId:
        """
        Register a listener to listen events from Notification Server

        :param listener_processor: The processor of the listener.
        :param event_keys: Keys of event for listening. If not set, it will listen all events.
        :param offset: The offset of the events to start listening.
        :return: The `ListenerRegistrationId` used to stop the listening.
        """
        pass

    @abc.abstractmethod
    def unregister_listener(self,
                            registration_id: ListenerRegistrationId):
        """
        Unregister the listener by id.
        """
        pass

    @abc.abstractmethod
    def list_events(self,
                    key: str = None,
                    namespace: str = None,
                    sender: str = None,
                    begin_offset: int = None,
                    end_offset: int = None) -> List[Event]:
        """
        List specific events in Notification Service.

        :param key: The key of the event to list.
        :param namespace: The namespace of the event to list.
        :param sender: The sender of the event to list.
        :param begin_offset: Offset of the events must be greater than this offset.
        :param end_offset: Offset of the events must be less than or equal to this offset.
        :return: The event list.
        """
        pass

    @abc.abstractmethod
    def time_to_offset(self, time: datetime) -> int:
        """
        Look up the biggest offset of the event whose create_time is
        less than or equal to the given timestamp.

        :return: The offset of the founded event.
        """
        pass

File: lib/notification_service/java/src/main/java/org/aiflow/notification/client/NotificationClient.java
package org.aiflow.notification.client;

import org.aiflow.notification.entity.EventKey;
import org.aiflow.notification.entity.EventMeta;

import java.util.List;

public abstract class NotificationClient {

    /**
     * Send the event to Notification Service.
     *
     * @param event The event to be sent.
     * @return Object of Event created in Notification Service.
     */
    protected abstract EventMeta sendEvent(EventMeta event) throws Exception;

    /**
     * Register a listener to start listen specific events in Notification Service.
     *
     * @param listenerProcessor Namespace of notification for listening.
     * @param eventKeys Keys of events for listening.
     * @param offset Start offset of notification for listening.
     */
    protected abstract ListenerRegistrationId registerListener(
            ListenerProcessor listenerProcessor, List<EventKey> eventKeys, Long offset);

    /**
     * * Unregister the listener by id.
     *
     * @param id The ListenerRegistrationId
     */
    protected abstract void unRegisterListener(ListenerRegistrationId id);

    /**
     * List specific events in Notification Service.
     *
     * @param event_name Name of events to be listed, null indicates not filter by name.
     * @param namespace Namespace of events to be listed, null indicates not filter by namespace.
     * @param eventType Type of events to be listed, null indicates not filter by type.
     * @param sender Sender of events to be listed, null indicates not filter by sender.
     * @param offset (Optional) The offset of events to start listing.
     * @return List of events in Notification Service.
     */
    protected abstract List<EventMeta> listEvents(
            String event_name, String namespace, String eventType, String sender, Long offset)
            throws Exception;

    /**
     * * Look up the offset of the first event whose create_time is bigger than the given timestamp
     *
     * @param timestamp The timestamp to look up.
     * @return The offset of the founded event.
     */
    protected abstract Long timeToOffset(Long timestamp) throws Exception;
}
